fluent/migrate/transforms.py

```diff
@@ -0,0 +1,281 @@
+# coding=utf8
+from __future__ import unicode_literals
+
+import fluent.syntax.ast as FTL
+
+
+def evaluate(ctx, node):
+    def eval_node(subnode):
+        if isinstance(subnode, Transform):
+            return subnode(ctx)
+        else:
+            return subnode
+
+    return node.traverse(eval_node)
+
+
+class Transform(FTL.Node):
+    def __call__(self, ctx):
+        raise NotImplementedError
+
+
+class SOURCE(Transform):
+    """Declare the source translation to be migrated with other transforms.
+
+    When evaluated `SOURCE` returns a simple string value.  All \\uXXXX from
+    the original translations are converted beforehand to the literal
+    characters they encode.
+
+    HTML entities are left unchanged for now because we can't know if they
+    should be converted to the characters they represent or not.  Consider the
+    following example in which `&amp;` could be replaced with the literal `&`:
+
+        Privacy &amp; History
+
+    vs. these two examples where the HTML encoding should be preserved:
+
+        Erreur&nbsp;!
+        Use /help &lt;command&gt; for more information.
+
+    """
+
+    # XXX Perhaps there's a strict subset of HTML entities which must or must
+    # not be replaced?
+
+    def __init__(self, path, key):
+        self.path = path
+        self.key = key
+
+    def __call__(self, ctx):
+        return ctx.get_source(self.path, self.key)
+
+
+class LITERAL(Transform):
+    """Create a Pattern with the literal text `value`.
+
+    This transform is used by `LITERAL_FROM` and can be used on its own with
+    `CONCAT`.
+    """
+
+    def __init__(self, value):
+        self.value = value
+
+    def __call__(self, ctx):
+        elements = [FTL.TextElement(self.value)]
+        return FTL.Pattern(elements)
+
+
+class LITERAL_FROM(SOURCE):
+    """Create a Pattern with the translation value from the given source."""
+
+    def __call__(self, ctx):
+        source = super(self.__class__, self).__call__(ctx)
+        return LITERAL(source)(ctx)
+
+
+class EXTERNAL(Transform):
```

Comment:
I'm considering a number of renames for the transforms, which I'll propose in a follow-up PR. Originally, all transforms except for `SOURCE` return a `Pattern` which makes them suitable for using as values of messages. Also `CONCAT` is implemented such that it can unwrap the expression from the `Pattern` and fold more than one transform into a single `Pattern`.

With a simple change to `CONCAT` it should be possible to do the same conditionally: i.e. unwrap `Patterns` when needed while keeping normal expressions unchanged. I think we'll end up with `EXTERNAL_EXPR` for selectors and replacements and `LITERAL_PATTERN`, `PLURAL_PATTERN` and `CONCAT_PATTERNS` for dealing with patterns.